man/man1/gravsim0.1
.TH GRAVSIM0 1NEMO "16 JUL 1991"
.SH NAME
gravsim0 \- NEMO frontend for the GRAVSIM family of programs
.SH SYNOPSIS
\fBgravsim0\fP [parameter=value]
.SH DESCRIPTION
\fIgravsim0\fP is a NEMO frontend to run any of the GRAVSIM family of
programs, such that NEMO datafile are given, converted on the fly to
GRAVSIM format, the specified GRAVSIM integrator is run, and output
files in GRAVSIM format are converted back to NEMO format.
.PP 
This program works by writing a C-shell script \fBgravsim\fP\fIpid\fP, 
where \fIpid\fP is the process ID (see \fIgetpid(3)\fP). This script
is then executed via \fIexecl(3)\fP, and kept in the current working
directory.
.PP
Since this program uses the following NEMO programs, they need
to have been installed: \fIstoa, 205-to-gdff, gdff-editor, gravsim*, 
gdff-to-205\fP and \fIatos\fP.
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBin=\fP
Snapshot input filename. No default.
.TP 20
\fBout=\fP
Snapshot output filename. No default.
.TP 20
\fBprog=\fP
GRAVSIM Program to run. [Default: \fBgravsim\fP].
.TP 20
\fBclean=\fP
Clean up temp files? [Default: \fBt\fP].
............................. remaining for GRAVSIM
.TP 20
\fBtitle=\fP
Title for the run    [Default: not used].
.TP 20
\fBprocessors=\fP
Number of processors that can be used  [Default: \fB1\fP].
.TP 20
\fBcells=\fP
Number of cells to be allocated  
.TP 20
\fBsplit-number=\fP
...       
.TP 20
\fBstart-time=\fP
Start time for run    [Default: \fB0.0\fP].
.TP 20
\fBend-time=\fP
Stop time for run    [Default: \fB2.0\fP].
.TP 20
\fBeps=\fP
Softening length      [Default: \fB0.025\fP].
.TP 20
\fBtol=\fP
Tolerance       [Default: \fB1.0\fP]
.TP 20
\fBdata-step=\fP
Major Output of data    [Default: \fB0.25\fP].
.TP 20
\fBreport-step=\fP
Minor Output of energy etc.   [Default: \fB0.03125\fP].
.TP 20
\fBg-value=\fP
Gravitational constant      [Default: \fB1.0\fP].
.TP 20
\fBtime-step=\fP
Integration time step     [Default: \fB0.03125\fP].
.SH SEE ALSO
gravsim(8GRAVSIM)
.SH AUTHOR
Peter Teuben, Mark Bellon
.SH BUGS
This program has not been tested extensively, it also serves as
another example to bind existing N-body integrators into NEMO.
The Aarseth NBODY family of N-body integrators shows another method
of doing the same job.
.SH FILES
.nf
.ta +2.5i
usr/mbellon/gravsim/	Source tree for GRAVSIM package (env.var. GRAVSIM)
usr/pjt/nbody      	gravsim0.c 
.fi
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
16-Jul-91	V0.0 document Created NEMO
.fi


src/nbody/evolve/gravsim/gravsim0.c
/*
 *  GRAVSIM0:   Nemo driver to the GRAVSIM family
 *
 *	Writes a script that:
 *	Converts input snapshot(5) via the "205" ascii format
 *	to gdff, adds the appropriate sections to the gdff
 *	file via standard NEMO keywords, runs the gravsim
 *	program, and converts the output datafile back to
 *	205, which is turned into output snapshot(5) again....
 *	The script is then executed via execl().
 *
 *	The following executables need to be present in the
 *	searchpath:
 *	   stoa, 205-to-gdff, gdff-editor, gravsim*, gdff-to-205 and atos
 *	   
 *
 * History:
 *	28-may-91	Created		PJT
 */
#include <stdinc.h>
#include <getparam.h>

string defv[] = {
/* NEMO parameters */
  "in=???\n                         Snapshot input filename",
  "out=???\n                        Snapshot output filename",
  "prog=gravsim\n                   GRAVSIM Program to run",
  "clean=t\n			    Clean up temp files?\n\
            Remaining parameters are for GRAVSIM family",
/* GRAVSIM prameters */
  "title=\n	                    Title for the run",
  "processors=1\n                   Number of processors that can be used",
  "cells=\n                         Number of cells to be allocated",
/*  "phi-and-acceleration-output=f\n  Extra output for phi/acc?",   */
  "split-number=\n                  ...",
  "start-time=0.0\n                 Start time for run",
  "end-time=2.0\n                   Stop time for run",
  "eps=0.05\n                       Softening length",
  "tol=1\n                          Tolerance",
  "data-step=0.25\n                 Major Output of data",
  "report-step=0.03125\n            Minor Output of energy etc.",
  "g-value=1\n                      Gravitational constant",
  "time-step=0.03125\n              Integration time step",
  "VERSION=1\n                      28-may-91 PJT",
  NULL,
};

string usage = "NEMO driver for gravsim N-body integrator";

#define NEMOPAR 3       /* define how many are NEMO pars, and GRAVSIM pars */

void
nemo_main()
{
    stream str;
    char tempname[100], *iname, *oname, *pname, *cp, key[20];
    int i, pid;

    iname = getparam("in");
    str = stropen(iname,"r");           /* open - to test if present */
    strclose(str);                      /* and close */
    pname = getparam("prog");
    if (*pname == NULL) error("Need prog= name");
    oname = getparam("out");

    /* make script */
    pid = getpid();
    sprintf(tempname,"gravsim%d",pid);
    dprintf(1,"[Writing temporary shell script %s\n",tempname);
    str = stropen(tempname,"w");
    fprintf(str,"#! /bin/csh -f\n");
    fprintf(str,"# File created by %s - do not edit\n",getargv0());
    fprintf(str,"stoa %s %s.1\n",iname,tempname);
    fprintf(str,"205-to-gdff %s.1 %s.2\n",tempname,tempname);
    fprintf(str,"gdff-editor %s.2 << END_OF_EDIT >& /dev/null\n",tempname);

    for (i=NEMOPAR; cp=defv[i]; i++) { /* process all args - except first few */
        if (cp==NULL) break;
        if (*cp==NULL) break;
        if (strncmp(cp,"VERSION",7)==0) break;
        strncpy(key,cp,20);
        for(cp=key; *cp; cp++)
            if (*cp=='=') {
                *cp=NULL;
                break;
            }
        cp = getparam(key);
        if (*cp == NULL) continue;
        dprintf(1,"Adding %s=%s\n",key,cp);
        fprintf(str,"add %s %s\n",key,cp);
    }
    fprintf(str,"exit\n");
    fprintf(str,"END_OF_EDIT\n");
    fprintf(str,"%s %s.2\n",pname,tempname);
    fprintf(str,"gdff-to-205 %s.2 %s.3\n",tempname,tempname);
    fprintf(str,"atos %s.3 %s\n",tempname,oname);
    if (getbparam("clean"))
        fprintf(str,"rm -f %s %s.1 %s.2 %s.3\n",
                        tempname,tempname,tempname,tempname);
    strclose(str);

    dprintf(0,"[Created and executing script %s]\n",tempname);
    execl("/bin/csh","csh",tempname,NULL);
    error("execl: %s\n",tempname);
}
